crates/services/p2p/src/heartbeat/handler.rs

```diff
@@ -0,0 +1,330 @@
+                Poll::Ready(Ok((stream, block_height))) => {
+                    // start waiting for the next `BlockHeight`
+                    self.inbound = Some(receive_block_height(stream).boxed());
+
+                    // report newly received `BlockHeight` to the Behaviour
+                    return Poll::Ready(ConnectionHandlerEvent::Custom(
+                        HeartbeatOutEvent::BlockHeight(block_height),
+                    ))
+                }
+                _ => {}
+            }
+        }
+
+        loop {
+            if self.failure_count >= self.config.max_failures.into() {
+                // Request from `Swarm` to close the faulty connection
+                return Poll::Ready(ConnectionHandlerEvent::Close(
+                    HeartbeatFailure::Timeout,
+                ))
+            }
+
+            match self.outbound.take() {
+                Some(OutboundState::RequestingBlockHeight { requested, stream }) => {
+                    self.outbound = Some(OutboundState::RequestingBlockHeight {
+                        stream,
+                        requested: true,
+                    });
+
+                    if !requested {
+                        return Poll::Ready(ConnectionHandlerEvent::Custom(
+                            HeartbeatOutEvent::RequestBlockHeight,
+                        ))
+                    }
+
+                    break
+                }
+                Some(OutboundState::SendingBlockHeight(mut outbound_block_height)) => {
+                    match outbound_block_height.poll_unpin(cx) {
+                        Poll::Pending => {
+                            if self.timer.poll_unpin(cx).is_ready() {
+                                // Time for successfull send expired!
+                                self.failure_count += 1;
+                                debug!(target: "fuel-libp2p", "Sending Heartbeat timed out, this is {} time it failed with this connection", self.failure_count);
+                            } else {
+                                self.outbound = Some(OutboundState::SendingBlockHeight(
+                                    outbound_block_height,
+                                ));
+                                break
+                            }
+                        }
+                        Poll::Ready(Ok(stream)) => {
+                            // reset failure count
+                            self.failure_count = 0;
+                            // start new idle timeout until next request & send
+                            self.timer.reset(self.config.idle_timeout);
+                            self.outbound = Some(OutboundState::Idle(stream));
+                        }
+                        Poll::Ready(Err(_)) => {
+                            self.failure_count += 1;
+                            debug!(target: "fuel-libp2p", "Sending Heartbeat failed, this is {} time it failed with this connection", self.failure_count);
```

Comment:
Consider using [ordinal](https://docs.rs/ordinal/latest/ordinal/) spelling or something else like:

```suggestion
                            debug!(target: "fuel-libp2p", "Sending Heartbeat failed, failure {}/{} for this connection", self.failure_count,  self.config.max_failures);
```